lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,78 @@
+use std::{fmt::Debug, ops::Deref};
+
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`CompiledModuleCache::load()`] will
+/// be called more often than [`CompiledModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait ModuleCache: Debug {
+    async fn load(&self, key: &str, engine: &Engine) -> Result<Module, CacheError>;
```

Comment:
As mentioned, I don't think this should be a generic string, but the sha256 hash of the wasm code.

An implementation should decide how to map a hash into a key, based on the engine.